### PR TITLE
feat(sqlite): migrate query-interface to typescript

### PR DESCRIPTION
**Pull request by lohart13**
## Pull Request Checklist

- [x] Have you added new tests to prevent regressions?
- [ ] If a documentation update is necessary, have you opened a PR to [the documentation repository](https://github.com/sequelize/website/)? 
- [x] Did you update the typescript typings accordingly (if applicable)?
- [x] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Does the name of your PR follow [our conventions](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md#6-commit-your-modifications)?

## Description Of Change

This PR add a new internal query-interface method `executeQueriesSequentially` which mean you can execute an array of sql strings for dialects which do not support multiple queries in a single string.

Also, all the query-interface methods for SQLite have been migrated to typescript.

## Todo:
- [ ] What tests are required for the new method.

### CODE REVIEW

File: packages/core/src/dialects/abstract/query-interface.d.ts
Comment (ephys) [inline packages/core/src/dialects/abstract/query-interface.d.ts:635]: Based on how our other methods are named, this should just be named `describeColumn`

That method already exists though, so a simple TODO is enough for this PR

File: packages/core/src/dialects/abstract/query-interface.d.ts
Comment (ephys) [inline packages/core/src/dialects/abstract/query-interface.d.ts:299]: this only exists in sqlite. It could be a method in `SqliteQueryInterfaceInternal` instead

File: packages/core/src/dialects/sqlite/query-interface.ts
Comment (ephys) [inline packages/core/src/dialects/sqlite/query-interface.ts:49]: This `AbstractQueryInterfaceInternal` instance should be passed upwards, otherwise a separate instance is going to be created

See how `PostgresQueryInterfaceTypeScript` does it

File: packages/core/src/dialects/sqlite/query-interface.ts
Comment (ephys) [inline packages/core/src/dialects/sqlite/query-interface.ts:53]: why clone the options if they're not modified?

File: packages/core/src/dialects/sqlite/query-interface.ts
Comment (ephys) [inline packages/core/src/dialects/sqlite/query-interface.ts:55]: prefer using EMPTY_ARRAY instead of creating a new array

File: packages/core/src/dialects/sqlite/query-interface.ts
Comment (ephys) [inline packages/core/src/dialects/sqlite/query-interface.ts:91]: This is not caused by your PR, but why would we include these properties in sqlite, but not the other dialects? The different dialects should be aligned, and in my opinion sqlite should align itself with the other dialects

Could you add a TODO?

File: packages/core/src/dialects/sqlite/query-interface.ts
Comment (ephys) [inline packages/core/src/dialects/sqlite/query-interface.ts:481]: should be moved to `SqliteQueryInterfaceInternal`

File: packages/core/src/dialects/sqlite/query-interface.ts
Comment (ephys) [inline packages/core/src/dialects/sqlite/query-interface.ts:386]: can you specify the return type of each of these methods explicitly?

File: packages/core/src/dialects/sqlite/query-interface.ts
Comment (ephys) [inline packages/core/src/dialects/sqlite/query-interface.ts:488]: prefer EMPTY_OBJECT over `{}`

**Review by ephys — DISMISSED** [review]

File: packages/core/src/dialects/abstract/query-interface.d.ts
Comment (lohart13) [reply packages/core/src/dialects/abstract/query-interface.d.ts:299]: `_dropAllTables` is used but the `dropAllTables` method in the `AbstractQueryInterface` to loop over the tables and drop them, it's not just used by SQLite.

This function could be merged into the `dropAllTables` method as this is the only use of it.

I can add a TODO if you want or make the change in this PR.

File: packages/core/src/dialects/abstract/query-interface.d.ts
Comment (lohart13) [reply packages/core/src/dialects/abstract/query-interface.d.ts:635]: Done

File: packages/core/src/dialects/sqlite/query-interface.ts
Comment (lohart13) [reply packages/core/src/dialects/sqlite/query-interface.ts]: I've updated this now and it uses the `SqliteQueryInterfaceInternal` as the `alterTableInternal` has been moved to this file.

File: packages/core/src/dialects/sqlite/query-interface.ts
Comment (lohart13) [reply packages/core/src/dialects/sqlite/query-interface.ts]: I've removed the unnecessary spreads now.

File: packages/core/src/dialects/sqlite/query-interface.ts
Comment (lohart13) [reply packages/core/src/dialects/sqlite/query-interface.ts]: Done.

File: packages/core/src/dialects/sqlite/query-interface.ts
Comment (lohart13) [reply packages/core/src/dialects/sqlite/query-interface.ts:81]: The reason for this is SQLite's limited ALTER TABLE function and it needs unique and foreign keys in the `ColumnDescription` so that they aren't overridden or removed. I've added comments in the code.

File: packages/core/src/dialects/sqlite/query-interface.ts
Comment (lohart13) [reply packages/core/src/dialects/sqlite/query-interface.ts]: Done

File: packages/core/src/dialects/sqlite/query-interface.ts
Comment (lohart13) [reply packages/core/src/dialects/sqlite/query-interface.ts]: Updated now.

File: packages/core/src/dialects/sqlite/query-interface.ts
Comment (lohart13) [reply packages/core/src/dialects/sqlite/query-interface.ts]: I updated `withSqliteForeignKeysOff` to accepted `QueryRawOptions` or `undefined` so the `{}` isn't required.

File: packages/core/src/dialects/abstract/query-interface.d.ts
Comment (lohart13) [reply packages/core/src/dialects/abstract/query-interface.d.ts]: I've submitted PR #16452 which merges the `dropAllTables` and `_dropAllTables` queries

**Comment by WikiRik** [comment]
I'm putting this to draft for until #16452 has been merged

**Review by WikiRik — APPROVED** [review]

File: packages/core/src/dialects/sqlite/query-interface.ts
Comment (lohart13) [reply packages/core/src/dialects/sqlite/query-interface.ts]: Further comments added now.
@ephys LMK if you are happy with these.

**Review by WikiRik — DISMISSED** [review]

**Review by WikiRik — APPROVED** [review]

**Comment by WikiRik** [comment]
Merging this to unblock https://github.com/sequelize/sequelize/pull/15934

**Pull request merged**
